lib/es/client.rb

```diff
@@ -72,6 +72,54 @@ def bulk(payload = {})
       end
     end
 
+    # Perform a search query with pagination and return a formatted response.
+    # ES paginates search results using a combination of `sort` and `search_after`.
+    # We repeat search queries until the response is empty, which is how we know pagination is complete.
+    def paginated_search(index_name, query) # rubocop:disable Metrics/MethodLength
+      results = []
+
+      loop do
+        retries = 0
+        begin
+          response = search(index: [index_name], body: query)
+          hits = response['hits']['hits']
+          # If hits is empty, no need to format results or continue pagination
+          return results if hits.empty?
+
+          results.push(*hits)
+          # Set `search_after` param for next paginated search call using the last hit's `sort` value
+          query['search_after'] = hits.last['sort']
+        rescue StandardError => e
+          retries += 1
+          if retries <= MAX_RETRIES
+            @system_logger.debug("Search attempt #{retries} failed. Retrying...")
+            sleep((2**retries).seconds) && retry
+          else
+            @system_logger.warn("Search failed after #{retries} attempts. #{e.message}")
+            raise e
+          end
+        end
+      end
+    end
+
+    def delete_by_query(index:, body:, refresh: true)
```

Comment:
> assuming it's "delete everything that still hasn't got an updated last_crawled_at"

It's not this, though. The query is deleting by specific IDs. As we run the purge crawl, if a crawl result is a success we remove the ID from `@purge_backlog`. At the end of the purge crawl, all remaining IDs in the backlog are deleted.

Your method could be a good improvement to save memory. I think it is the way it is because I initially wasn't updating docs found during the purge crawl, so the `last_crawled_at` wouldn't have been a good reference.

Reply:
I've changed it to use the same query as the search query (`lt` range based on` last_crawled_at`)